### DISCUSSION TITLE
Endpoint visible for container

### DISCUSSION THREAD

**Discussion by rosieks** [microsoft/aspire#5954] [Q&A] [answered]
I strugle with getting correct endpoint. I have the following setup:
* Keycloak (hosted on docker)
* Grafana (hosted on docker)
* My API

I have plugin for Grafana that calls keycloak to get bearer token to access my api. To get keycloak adress I call `keycloak.GetEndpoint("http").Url`. Unfortunatelly it returns `localhost:8080` so when grafana calls that url it gets connection refused. How can I get correct url (`host.docker.internal`) that will be visible from keycloak?

**Comment by davidfowl** [comment]
Show your app host code. Currrrently container to container networking required some understanding of the APIs. That will improve in Aspire 9, but I can show you how to make it work.

**Reply by rosieks** [reply]
My code looks like that:

```csharp
// keycloak
var keycloak = builder.AddKeycloakContainer("keycloak");

// grafana
var grafana = builder.AddContainer("grafana", "grafana/grafana")
    .WithEndpoint(port: 3010, targetPort: 3000, scheme: "http", name: "http")

// api
var api = builder.AddProject<Projects.ProvisioningApi>()
    .WithEnvironment(ctx => {
        ctx.EnvironmentVariables["Grafana__Url"] = grafana.GetEndpoint("http").Url;
        // ctx.EnvironmentVariables["Keycloak__AuthServerUrl"] = $"http://host.docker.internal:{keycloak.GetEndpoint("http").Port}";
        ctx.EnvironmentVariables["Keycloak__AuthServerUrl"] = keycloak.GetEndpoint("http").Url;
    });
```

In my ProvisioningApi I call grafana and setup some plugin. One of the step is providing URL to keycloak so grafan could call it to get token. To workaround that I just hardcoded *host.docker.internal* but it would be nice to have something more sophisticated

**Reply by Inx51** [reply]
Same here..
Trying to setup the eShop-sample with flagd for feature flagging/toggeling.. however If I simply call GetEndpoint("http") it returns an endpoint pointing to http://localhost:8013, which I assume would make more sense if this would be the internal host..
Also.. for me I end up with an exception if I call .GetEndpoint("http").Url (or anything that actually requires the method to run) 

```csharp
var flagd = builder.AddContainer("flagd", "ghcr.io/open-feature/flagd", "latest")
    .WithBindMount("X:/flagd", "/flagd")
    .WithArgs("start", "-f", "file:/flagd/demo.flagd.json")
    .WithEndpoint(scheme: "http", name: "http", port: 8013, targetPort: 8013);

.....

var webApp = builder.AddProject<Projects.WebApp>("webapp", launchProfileName)
    .WithExternalHttpEndpoints()
    .WithReference(basketApi)
    .WithReference(catalogApi)
    .WithReference(orderingApi)
    .WithReference(rabbitMq)
    //.WithEnvironment("Flagd__Endpoint", flagd.GetEndpoint("http"))  ... this sort of works, but ends up with the host beeing localhost and not the internal docker containers host.
    //.WithEnvironment("Flagd__Endpoint", flagd.GetEndpoint("http").Url)  ... this fails with the exception provided below..
    .WithEnvironment("IdentityUrl", identityEndpoint);
```
Exception:
Unhandled exception. System.InvalidOperationException: The endpoint `http` is not allocated for the resource `flagd`.
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_AllocatedEndpoint() in XXXXXXXXXXXXXXXXX/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 105
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_Url() in XXXXXXXXXXXXXXXX/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 102
   at Program.<Main>$(String[] args) in XXXXXXXXXXXXXXXXXXXXXXXXXXXX\eShop.AppHost\Program.cs:line 72

I guess Im missing something here..

**Reply by davidfowl** [reply, marked as answer]
Sorry for taking so long to respond. This is a tricky topic as we attempt to do some magic under the covers to make sure things "just work" in the default case of passing an endpoint reference or passing a connection string via WithReference.

When you get an `EndpointReference` (via `resource.GetEndpoint`), it has no idea what the URL is or is going to be until the app host starts up. That's why you see exceptions about the endpoint not being allocated when you try to resolve the Url during startup. This is why there's an environment variable callback. It will run after the endpoints have been allocated and you can access various properties.

When we detect a container to host reference when resolving value, we replace this value with the container host (docker.host.internal). To avoid this translation you need to resolve the endpoint reference in the right context (in RunMode and in the environment variable callback or when the `AfterEndpointsAllocated` event fires).

@rosieks For your "delegated scenario", you're trying to give your provisioner keycloack's url from the POV of the grafana resource, you have to manually replace the host name in run mode. If you want this to work in publish mode then need to handle both cases.

```C#
// keycloak
var keycloak = builder.AddKeycloakContainer("keycloak");

// grafana
var grafana = builder.AddContainer("grafana", "grafana/grafana")
    .WithEndpoint(port: 3010, targetPort: 3000, scheme: "http", name: "http")

// api
var api = builder.AddProject<Projects.ProvisioningApi>()
.WithEnvironment(ctx =>
{
    if (ctx.ExecutionContext.IsRunMode)
    {
        ctx.EnvironmentVariables["Grafana__Url"] = grafana.GetEndpoint("http");

        // We want to tell grafana what the keycloak URL is, but we need to use the container host
        // so that it will resolve correctly from the container

        var keyCloakEndpoint = keycloak.GetEndpoint("http");
        var keycloakUrl = new UriBuilder(keyCloakEndpoint.Url)
        {
            Host = keyCloakEndpoint.ContainerHost
        };

        ctx.EnvironmentVariables["Keycloak__AuthServerUrl"] = keycloakUrl.ToString();
    }
    else
    {
        // In publish mode let the endpoint resolver handle the URL
        ctx.EnvironmentVariables["Grafana__Url"] = grafana.GetEndpoint("http");
        ctx.EnvironmentVariables["Keycloak__AuthServerUrl"] = keycloak.GetEndpoint("http");
    }
});
```

We're looking at exposing an API to resolve an endpoint reference in the right networking context.

**Reply by JobaDiniz** [reply]
>We're looking at exposing an API to resolve an endpoint reference in the right networking context.

is this still required or has this been implemented officially?

**Reply by davidfowl** [reply]
https://github.com/dotnet/aspire/issues/7153

**Reply by MGREMY** [reply]
Yupp feels tricky, could be really cool to have an official solution for this one !
